src/mongocrypt-ctx.c

```diff
@@ -172,6 +172,78 @@ mongocrypt_ctx_setopt_key_alt_name (mongocrypt_ctx_t *ctx,
 }
 
 
+bool
+mongocrypt_ctx_setopt_key_material (mongocrypt_ctx_t *ctx,
+                                    mongocrypt_binary_t *key_material)
+{
+   bson_t as_bson;
+   bson_iter_t iter;
+   const char *key;
+   _mongocrypt_buffer_t buffer;
+
+   if (!ctx) {
+      return false;
+   }
+
+   if (ctx->initialized) {
+      return _mongocrypt_ctx_fail_w_msg (ctx, "cannot set options after init");
+   }
+
+   if (ctx->opts.key_material.owned) {
+      return _mongocrypt_ctx_fail_w_msg (ctx, "keyMaterial already set");
+   }
+
+   if (ctx->state == MONGOCRYPT_CTX_ERROR) {
+      return false;
+   }
+
+   if (!key_material || !key_material->data) {
+      return _mongocrypt_ctx_fail_w_msg (ctx, "option must be non-NULL");
+   }
+
+   if (!_mongocrypt_binary_to_bson (key_material, &as_bson)) {
+      return _mongocrypt_ctx_fail_w_msg (ctx,
+                                         "invalid keyMaterial bson object");
+   }
+
+   if (!bson_iter_init (&iter, &as_bson) || !bson_iter_next (&iter)) {
```

Comment:
I was going to suggest using `_mongocrypt_parse_required_binary` to simplify parsing the BSON binary. `_mongocrypt_parse_required_binary` is used for `gcp.privateKey` and `local.key`. But a test case you wrote uncovered MONGOCRYPT-380.

I suggest adding a note to address this after MONGOCRYPT-380 is resolved. `_mongocrypt_parse_required_binary` accepts both BSON binary and BSON UTF-8 representing base64.

```suggestion
   /* TODO: Once MONGOCRYPT-380 is resolved, use _mongocrypt_parse_required_binary to parse "keyMaterial" */
   if (!bson_iter_init (&iter, &as_bson) || !bson_iter_next (&iter)) {
```

Reply:
Noted. Will add corresponding TODO comment.